Outlines Regression ADM

jadie1 opened the pull request:
Implementation of the multi-KDMA/regression ADM using outlines
To test, call `run_align_system` with `adm=outlines_regression_aligned`

The ADM is implemented in [outlines_regression_adm.py](https://github.com/ITM-Kitware/align-system/blob/dev/outlines-regression/align_system/algorithms/outlines_regression_adm.py). It inherits functionality from [outlines_adm.py](https://github.com/ITM-Kitware/align-system/blob/dev/outlines-regression/align_system/algorithms/outlines_adm.py) but the `top_level_choose_action()` function is overwritten to do KDMA score prediction. Prompts and json schemas for this have been added to [outlines_prompts.py](https://github.com/ITM-Kitware/align-system/blob/dev/outlines-regression/align_system/prompt_engineering/outlines_prompts.py). I also added a [kdma_descriptions.yml](https://github.com/ITM-Kitware/align-system/blob/dev/outlines-regression/align_system/prompt_engineering/kdma_descriptions.yml) file to keep track of the official KDMA descriptions for various prompts.

jadie1 commented:
One thing I am unsure of is the follow-up prompts based on the selected choice in [choose_action()](https://github.com/ITM-Kitware/align-system/blob/main/align_system/algorithms/outlines_adm.py#L240). Right now [outlines_regression_adm.py](https://github.com/ITM-Kitware/align-system/blob/dev/outlines-regression/align_system/algorithms/outlines_regression_adm.py) inherits this function directly from [outlines_adm.py](https://github.com/ITM-Kitware/align-system/blob/dev/outlines-regression/align_system/algorithms/outlines_adm.py) and the  `top_level_choose_action()` function returns a simplified version of the dialog to add follow-up clarification prompts to if needed. Let me know if this reasonable or requires more thought.

dmjoy commented on align_system/algorithms/outlines_regression_adm.py:
```diff
@@ -0,0 +1,312 @@
+log = logging.getLogger(__name__)
+JSON_HIGHLIGHTER = JSONHighlighter()
+
+# TODO - make this configurable? 
+KDMA_DESCRIPTIONS_FILE_PATH = os.path.join(
+    pathlib.Path(__file__).parent.absolute(), '..',
+    'prompt_engineering/kdma_descriptions.yml')
```
Yes that would be ideal, but let's save for a follow-on MR

dmjoy commented on align_system/algorithms/outlines_regression_adm.py:
```diff
@@ -0,0 +1,312 @@
+
+        # Sample multiple predictions
+        for _ in range(num_samples):
+            # Shuffle
+            if shuffle_choices:
+                shuffled_choices = random.sample(choices, len(choices))
+            else:
+                shuffled_choices = choices
```
Question, if we're prompt the LLM about once specific choice at a time, would shuffling them even make a difference?

jadie1 replied:
Great point, removed in [f8683f8](https://github.com/ITM-Kitware/align-system/pull/62/commits/f8683f87e21710ca52c4c68f3ae18fc15f8f94e4)

dmjoy commented on align_system/algorithms/outlines_regression_adm.py:
```diff
@@ -0,0 +1,312 @@
+                    'reasonings':[]     # List to add sampled kdma score reasonings to
+                }
+
+        # Sample multiple predictions
+        for _ in range(num_samples):
```
Outlines can handle this for us by passing in multiple prompts at a time in a list, unless there's some other compelling reason to use a loop here (we don't have to refactor this now for this MR, but a potential optimization)

jadie1 replied:
Refactored in [3a6f9f6](https://github.com/ITM-Kitware/align-system/pull/62/commits/3a6f9f61b1a557598f8e2cf0d58b8597a3e9a765)

dmjoy commented on align_system/algorithms/outlines_regression_adm.py:
```diff
@@ -0,0 +1,312 @@
+        if distribution_matching == 'average':
+            # Averages over predicted score samples and selects choice with minimum MSE to target
+            selected_choice, first_justification = self.average_distribution_matching(predicted_kdma_values, target_kdmas)
+            # Currently returning the reasoning associated with the first sample for the selected choice
```
Good comment

dmjoy commented on align_system/prompt_engineering/kdma_descriptions.yml:
I like that you've gone ahead and added descriptions for the new KDMAs, was this a collaborative effort with Emily et. al?  If not you'll want to share these around (perhaps in the ITM group chat) so that we don't duplicate work to generate them

jadie1 replied:
Good idea, I shared it

dmjoy reviewed (commented):
Overall this looks really good, one potentially major tweak here would be to remove the option of choice shuffling, since I don't think it affects the outcome (we can discuss more if needed), and secondary to that would be making use of outlines' built in functionality for multiple sampling.

jadie1 commented on align_system/algorithms/outlines_regression_adm.py:
```diff
@@ -0,0 +1,312 @@
+            whitespace_pattern=r"[ ]?")
+
+        kdma_dialog_texts = [self.dialog_to_prompt(d) for d in
+            itertools.chain(kdma_dialogs)]
```
Addressed in [f79700a](https://github.com/ITM-Kitware/align-system/pull/62/commits/f79700a26a180c253aa8b2e5841ada83d8916257)

jadie1 commented on align_system/algorithms/outlines_regression_adm.py:
```diff
@@ -0,0 +1,312 @@
+        for choice in choices:
+            predict_outcome_prompt = outcome_prediction_prompt(scenario_description, choice)
+            outcome_dialogs.append([{'role': 'system', 'content': outcomes_sys_prompt},
+                        {'role': 'user', 'content': predict_outcome_prompt}])
```
Addressed in [f79700a](https://github.com/ITM-Kitware/align-system/pull/62/commits/f79700a26a180c253aa8b2e5841ada83d8916257)

jadie1 commented on align_system/algorithms/outlines_regression_adm.py:
```diff
@@ -0,0 +1,312 @@
+    # Predicts kdma scores associated with each choice
+    # Outputs a list of ADM responses and a corresponding list of response keys:
+    #   kdma_score_responses = [{score:int, reasoning:str}, ...]
+    #   reponse_keys = [{kdma:str, choice:str}, ...]
```
Addressed in [f79700a](https://github.com/ITM-Kitware/align-system/pull/62/commits/f79700a26a180c253aa8b2e5841ada83d8916257)

jadie1 commented on align_system/algorithms/outlines_regression_adm.py:
```diff
@@ -0,0 +1,312 @@
+                 extra={"markup": True})
+        log.info(kdma_dialog_texts[0])
+
+        # List of {score:int, reasoning:str}
```
Thanks for pointing this out. The original implementation did have the LLM provide reasoning first so I'll leave this order for now, but this will be good to reconsider when we think about ways to improve performance

jadie1 commented:
@dmjoy I've addressed all your comments now and refactored it to use outlines sampling functionality. It now predicts outcomes for all samples first then predicts kdma scores for all samples. In this way the same generators are used for all samples. I've also added a `generator_batch_size` parameter which controls how many prompts are sent at a time to avoid OOM errors.
